rest-client/src/main/java/org/sagebionetworks/bridge/rest/ApiClientProvider.java

```diff
@@ -74,54 +180,49 @@ public UserSessionInfoProvider getUserSessionInfoProvider(SignIn signIn) {
         UserSessionInfoProvider sessionProvider = null;
         synchronized(this) {
-            sessionProvider = sessionProviders.get(internalSignIn);
+            sessionProvider = signInSessionProviders.get(internalSignIn);
             if (sessionProvider == null) {
                 sessionProvider = new UserSessionInfoProvider(getClient(AuthenticationApi.class), internalSignIn);
-                sessionProviders.put(internalSignIn, sessionProvider);
+                emailSessionProviders.put(internalSignIn.getEmail(), sessionProvider);
+                signInSessionProviders.put(internalSignIn, sessionProvider);
             }
+            return getClient(sessionProvider, service);
         }
-        
-        UserSessionInterceptor sessionInterceptor = new UserSessionInterceptor(sessionProvider);
-        AuthenticationHandler authenticationHandler = new AuthenticationHandler(sessionProvider);
+    }
 
-        OkHttpClient.Builder httpClientBuilder = getHttpClientBuilder(sessionInterceptor, authenticationHandler);
-        httpClientBuilder.authenticator(authenticationHandler);
-        Retrofit retrofit = getRetrofit(httpClientBuilder.build());
-        T authenticatingClient = retrofit.create(service);
+    /**
+     *
+     * @param service Class representing the service
+     * @param email email for the user, or null for an unauthenticated client
+     * @param <T> One of the Api classes in the org.sagebionetworks.bridge.rest.api package.
+     * @return service client that is authenticated with the user's credentials
+     */
+    public <T> T getClient(Class<T> service, String email) {
+        checkNotNull(service);
+        checkNotNull(email);
 
-        return authenticatingClient;
+        return getClient(emailSessionProviders.get(email), service);
     }
-    
-    OkHttpClient.Builder getHttpClientBuilder(Interceptor... interceptors) {
-        OkHttpClient.Builder builder = new OkHttpClient.Builder()
-            .connectTimeout(2, TimeUnit.MINUTES)
-            .readTimeout(2, TimeUnit.MINUTES)
-            .writeTimeout(2, TimeUnit.MINUTES);
-        for (Interceptor interceptor : interceptors) {
-            builder.addInterceptor(interceptor);
-        }
-        return builder
-            .addInterceptor(new HeaderInterceptor(userAgent, acceptLanguage))
-            .addInterceptor(WARNING_INTERCEPTOR)
-            .addInterceptor(ERROR_INTERCEPTOR)
-            .addInterceptor(LOGGING_INTERCEPTOR);
-    }
-    
-    // allow test access to retrofit builder
-    Retrofit getRetrofit(OkHttpClient client) {
-        return getRetrofitBuilder()
-            .baseUrl(baseUrl)
-            .client(client)
-            .addConverterFactory(GsonConverterFactory.create(RestUtils.GSON))
-            .build();
+
+    public void setEmailUserSessionInfo(String studyId, String email, UserSessionInfo session) {
+        emailSessionProviders.putIfAbsent(email,
```

Comment:
Having this setter does seem awkward given the rest of the methods function like a factory. Perhaps another reason in favor of a builder pattern.